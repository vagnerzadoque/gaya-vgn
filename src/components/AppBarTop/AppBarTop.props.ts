type AppBarTopColor = 'default' | 'primary' | 'none' | 'inverse' | 'secondary'
type AppBarTopBehaviour = 'fixed' | 'scroll' | 'absolute'

export type AppBarProps = {
  /**
   * If `fixed` the AppBarTop is fixed leaving the content
   * scrolling from behind, when using `scroll` it hides
   * the AppBarTop when scrolling up and reveals when scrolling
   * down
   * @default fixed
   */
  behaviour?: AppBarTopBehaviour

  /**
   * Specify the content of your AppBarTop
   */
  children?: React.ReactNode

  /**
   * Override or extend the styles applied to the component
   */
  className?: string

  /**
   * Define de background color
   * @default default
   */
  color?: AppBarTopColor

  /**
   * If `false`, a box-shadow effect are disabled
   * @default true
   */

  elevation?: boolean

  /**
   * Define a semantic role for Assistive Technology
   */
  role?: string

  /**
   * Optional ID for testing
   */
  testID?: string
}

export type AppBarStyleOptions = AppBarProps & {
  hide: boolean
}
