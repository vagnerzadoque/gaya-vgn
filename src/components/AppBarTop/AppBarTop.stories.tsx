/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { fn } from '@storybook/test'
import TextField from '../TextField'
import { IconButton } from '../IconButton'
import { Icon } from '../Icon'
import { AppBarTop, AppBarProps } from '.'
import { Logo } from '../Logo'

const componentStatus = `
> The App Bar Top displays information and actions relating to the current screen.

## Properties
| Property                      | Values                                 |    Status           |
|---                            |                                     ---|                  ---|
| **variant(no prop)**                   | base/standard                 | ✅ Available        |
| **color**                     | default, primary, inverse, none        | ✅ Available        |
| **elevation**                 | true/false                             | ✅ Available        |
| **behaviour**                 | fixed, scroll, absolute                | ✅ Available        |
| **action**                    | left, right                            | ✅ (available but with alternative composition)        |
| **Content Position**          | (available but with alternative composition) | ✅ (available but with alternative composition)        |
| **Content Type**              | (available but with alternative composition) | ✅ (available but with alternative composition)        |
| **Content Proeminent**        | (available but with alternative composition) | ✅ (available but with alternative composition)        |


## Technical Usages Examples
`

const meta: Meta<AppBarProps> = {
  title: 'Components/AppBarTop',
  component: AppBarTop,
  parameters: {
    componentSubtitle: '',
    docs: { description: { component: componentStatus } }
  }
};

export default meta;
type Story = StoryObj<AppBarProps>;

export const Playground: Story = {
  args: {
    children: 'App Bar - Top',
    behaviour: 'absolute'
  },
  render: (args: AppBarProps) => (
    <div style={{
      display: 'flex',
      position: 'relative',
      gap: 8,
      flexDirection: 'column',
      minHeight: 920,
      width: '100%',
      // border: '1px solid red',
      overflowY: 'scroll'
    }}
    >
      <AppBarTop {...args} />
    </div>
  )
}

export const Color: Story = {
  args: {
    children: 'App Bar - Top',
    style: { position: 'static' } as React.CSSProperties
  },
  render: (args: AppBarProps) => (
    <div style={{ display: 'flex', gap: 8, flexDirection: 'column' }}>
      <AppBarTop {...args} color="default">Default</AppBarTop>
      <AppBarTop {...args} color="primary">Primary</AppBarTop>
      <AppBarTop {...args} color="secondary">Secondary</AppBarTop>
      <AppBarTop {...args} color="inverse">Inverse</AppBarTop>
      <AppBarTop {...args} color="none">None</AppBarTop>
    </div>
  )
}

export const Elevation: Story = {
  args: {
    children: 'App Bar - Top',
    style: { position: 'static' } as React.CSSProperties
  },
  render: (args: AppBarProps) => (
    <div style={{ display: 'flex', gap: 8, flexDirection: 'column' }}>
      <AppBarTop {...args}>Elevation</AppBarTop>
      <AppBarTop {...args} elevation={false}>No Elevation</AppBarTop>
    </div>
  )
}

export const WithLogo: Story = {
  args: {
    style: { position: 'static' } as React.CSSProperties
  },
  render: (args: AppBarProps) => (
    <AppBarTop {...args}>
      <div style={{ display: 'flex', alignItems: 'center', width: '100%' }}>
        <IconButton IconComponent={<Icon name="outlined-navigation-menu" color="highlight" />} ariaLabel="menu" onClick={fn()} />
        <div style={{ flexGrow: 1, justifyContent: 'center', display: 'flex' }}>
          <Logo size="hugeX" />
        </div>
        <>
          <IconButton IconComponent={<Icon name="outlined-action-search" color="highlight" />} ariaLabel="search" onClick={fn()} />
          <IconButton IconComponent={<Icon name="outlined-finance-bag" color="highlight" />} ariaLabel="profile" onClick={fn()} />
        </>
      </div>
    </AppBarTop>
  )
}
export const WithSearch: Story = {
  args: {
    style: { position: 'static' } as React.CSSProperties
  },
  render: (args: AppBarProps) => (
    <AppBarTop {...args}>
      <>
        <div style={{ justifyContent: 'center', display: 'flex' }}>
          <IconButton IconComponent={<Icon name="outlined-navigation-menu" color="highlight" />} ariaLabel="menu" onClick={fn()} />
        </div>
        <div style={{ justifyContent: 'center', display: 'flex' }}>
          <TextField
            ariaLabel="search"
            onClick={fn()}
            id="123"
            size="medium"
            placeholder="Search"
            action="icon"
            IconComponent={<IconButton IconComponent={<Icon name="outlined-action-search" color="highlight" />} ariaLabel="search" onClick={fn()} />}
          />
        </div>
        <div>
          <IconButton IconComponent={<Icon name="outlined-action-search" color="highlight" />} ariaLabel="search" onClick={fn()} />
          <IconButton IconComponent={<Icon name="outlined-finance-bag" color="highlight" />} ariaLabel="profile" onClick={fn()} />
        </div>
      </>
    </AppBarTop>
  )
}
