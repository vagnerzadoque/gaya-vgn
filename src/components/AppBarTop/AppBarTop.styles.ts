import { createUseStyles } from 'react-jss'
import type { Theme } from '@naturacosmeticos/natds-themes'
import type { AppBarStyleOptions } from './AppBarTop.props'

const getBackgroundColor = (theme: Theme) => ({ color }: AppBarStyleOptions) => {
  const backgroundColor = {
    default: theme.color.surface,
    primary: theme.color.primary,
    secondary: theme.color.secondary,
    none: 'transparent',
    inverse: theme.color.highEmphasis
  }

  return color && backgroundColor[color]
}

const getTextColor = (theme: Theme) => ({ color }: AppBarStyleOptions) => {
  const backgroundColor = {
    default: theme.color.onSurface,
    primary: theme.color.onPrimary,
    secondary: theme.color.onSecondary,
    none: theme.color.highlight,
    inverse: theme.color.surface
  }

  return color && backgroundColor[color]
}

const styles = createUseStyles((theme: Theme) => ({
  container: ({ behaviour, hide }: AppBarStyleOptions) => {
    const shared = {
      position: 'fixed',
      top: 0,
      left: 0,
      width: '100%',
      zIndex: 1100
    }

    if (behaviour === 'fixed') {
      return {
        ...shared
      }
    }
    if (behaviour === 'absolute') {
      return {
        ...shared,
        position: 'absolute'
      }
    }

    return {
      ...shared,
      transition: '0.3s ease-in-out',
      top: hide ? -1 * (theme.size.mediumX + theme.spacing.micro * 2) : 0
    }
  },
  appBar: {
    alignItems: 'center',
    display: 'flex',
    flex: 'auto',
    flexDirection: 'row',
    justifyContent: 'space-between',
    minHeight: theme.size.mediumX,
    padding: [theme.spacing.micro, theme.spacing.tiny],
    boxSizing: 'border-box',
    width: 'auto',
    backgroundColor: getBackgroundColor(theme),
    boxShadow: ({ elevation }: AppBarStyleOptions) => (elevation ? theme.elevation.tiny : 'none'),
    color: getTextColor(theme),
    fontFamily: [
      theme.appBarTop.title.primary.fontFamily,
      theme.appBarTop.title.fallback.fontFamily
    ],
    fontSize: theme.appBarTop.title.fontSize,
    fontWeight: theme.appBarTop.title.primary.fontWeight,
    letterSpacing: theme.appBarTop.title.letterSpacing,
    lineHeight: theme.appBarTop.title.lineHeight
  }
}))

export default styles
