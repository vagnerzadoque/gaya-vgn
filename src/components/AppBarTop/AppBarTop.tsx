import React, { useLayoutEffect, useState } from 'react'
import styles from './AppBarTop.styles'
import type { AppBarProps } from './AppBarTop.props'

const AppBarTop = React.forwardRef<HTMLElement, AppBarProps>(
  (
    {
      className = '',
      color = 'default',
      behaviour = 'fixed',
      elevation = true,
      testID,
      children,
      role,
      ...props
    },
    ref
  ) => {
    const [position, setPosition] = useState({
      hide: false,
      y: global.scrollY
    })

    useLayoutEffect(() => {
      const onScroll = () => {
        const currentY = global.scrollY

        if (currentY < position.y) {
          setPosition({ y: currentY, hide: false })
        } else if (currentY > position.y) {
          setPosition({ y: currentY, hide: true })
        }
      }

      document.addEventListener('scroll', onScroll)

      return () => {
        document.removeEventListener('scroll', onScroll)
      }
    }, [position.y])

    const { container, appBar } = styles({
      elevation,
      color,
      behaviour,
      hide: position.hide
    })

    return (
      <header ref={ref} className={`${className} ${container} AppBarTop-gaya`} {...props} role={role} data-testid={testID}>
        <div className={appBar}>
          {children}
        </div>
      </header>
    )
  }
)

export default AppBarTop
