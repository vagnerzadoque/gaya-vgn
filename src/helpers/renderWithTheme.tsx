import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import themes from '@naturacosmeticos/natds-themes'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component: JSX.Element) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider theme={themes.natura.light}>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
