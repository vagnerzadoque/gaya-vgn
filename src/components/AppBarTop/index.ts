import AppBarTop from './AppBarTop'

export { AppBarTop }
export { AppBarProps } from './AppBarTop.props'
export default AppBarTop
