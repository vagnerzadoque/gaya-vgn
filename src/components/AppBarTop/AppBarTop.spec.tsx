import React from 'react'
import AppBarTop from '.'
import renderWithTheme from '../../helpers/renderWithTheme'

describe('AppBarTop component', () => {
  it('should render correctly with default props', () => {
    const { styles, component } = renderWithTheme(<AppBarTop>app bar</AppBarTop>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly without elevation', () => {
    const { styles, component } = renderWithTheme(<AppBarTop elevation={false}>app bar</AppBarTop>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly with primary color variant', () => {
    const { styles, component } = renderWithTheme(<AppBarTop color="primary">app bar</AppBarTop>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly with inverse color variant', () => {
    const { styles, component } = renderWithTheme(<AppBarTop color="inverse">app bar</AppBarTop>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly with none color variant', () => {
    const { styles, component } = renderWithTheme(<AppBarTop color="none">app bar</AppBarTop>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
})
