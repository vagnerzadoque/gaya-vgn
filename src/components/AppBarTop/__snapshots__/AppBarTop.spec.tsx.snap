// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBarTop component should render correctly with default props 1`] = `
[
  ".container-0-2-1 {}
.appBar-0-2-2 {
  flex: auto;
  width: auto;
  display: flex;
  padding: 4px 8px;
  font-size: 20px;
  box-sizing: border-box;
  min-height: 56px;
  align-items: center;
  font-family: Roboto, Arial;
  font-weight: 500;
  line-height: 1.25;
  flex-direction: row;
  letter-spacing: 0.24px;
  justify-content: space-between;
}
.container-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  z-index: 1100;
  position: fixed;
}
.appBar-d1-0-2-4 {
  background-color: #ffffff;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14);
  color: #363636;
}",
  <div>
    <header
      class=" container-0-2-1 container-d0-0-2-3 AppBarTop-gaya"
    >
      <div
        class="appBar-0-2-2 appBar-d1-0-2-4"
      >
        app bar
      </div>
    </header>
  </div>,
]
`;

exports[`AppBarTop component should render correctly with inverse color variant 1`] = `
[
  ".container-0-2-1 {}
.appBar-0-2-2 {
  flex: auto;
  width: auto;
  display: flex;
  padding: 4px 8px;
  font-size: 20px;
  box-sizing: border-box;
  min-height: 56px;
  align-items: center;
  font-family: Roboto, Arial;
  font-weight: 500;
  line-height: 1.25;
  flex-direction: row;
  letter-spacing: 0.24px;
  justify-content: space-between;
}
.container-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  z-index: 1100;
  position: fixed;
}
.appBar-d1-0-2-4 {
  background-color: #363636;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14);
  color: #ffffff;
}",
  <div>
    <header
      class=" container-0-2-1 container-d0-0-2-3 AppBarTop-gaya"
    >
      <div
        class="appBar-0-2-2 appBar-d1-0-2-4"
      >
        app bar
      </div>
    </header>
  </div>,
]
`;

exports[`AppBarTop component should render correctly with none color variant 1`] = `
[
  ".container-0-2-1 {}
.appBar-0-2-2 {
  flex: auto;
  width: auto;
  display: flex;
  padding: 4px 8px;
  font-size: 20px;
  box-sizing: border-box;
  min-height: 56px;
  align-items: center;
  font-family: Roboto, Arial;
  font-weight: 500;
  line-height: 1.25;
  flex-direction: row;
  letter-spacing: 0.24px;
  justify-content: space-between;
}
.container-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  z-index: 1100;
  position: fixed;
}
.appBar-d1-0-2-4 {
  background-color: transparent;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14);
  color: #111111;
}",
  <div>
    <header
      class=" container-0-2-1 container-d0-0-2-3 AppBarTop-gaya"
    >
      <div
        class="appBar-0-2-2 appBar-d1-0-2-4"
      >
        app bar
      </div>
    </header>
  </div>,
]
`;

exports[`AppBarTop component should render correctly with primary color variant 1`] = `
[
  ".container-0-2-1 {}
.appBar-0-2-2 {
  flex: auto;
  width: auto;
  display: flex;
  padding: 4px 8px;
  font-size: 20px;
  box-sizing: border-box;
  min-height: 56px;
  align-items: center;
  font-family: Roboto, Arial;
  font-weight: 500;
  line-height: 1.25;
  flex-direction: row;
  letter-spacing: 0.24px;
  justify-content: space-between;
}
.container-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  z-index: 1100;
  position: fixed;
}
.appBar-d1-0-2-4 {
  background-color: #f4ab34;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14);
  color: #333333;
}",
  <div>
    <header
      class=" container-0-2-1 container-d0-0-2-3 AppBarTop-gaya"
    >
      <div
        class="appBar-0-2-2 appBar-d1-0-2-4"
      >
        app bar
      </div>
    </header>
  </div>,
]
`;

exports[`AppBarTop component should render correctly without elevation 1`] = `
[
  ".container-0-2-1 {}
.appBar-0-2-2 {
  flex: auto;
  width: auto;
  display: flex;
  padding: 4px 8px;
  font-size: 20px;
  box-sizing: border-box;
  min-height: 56px;
  align-items: center;
  font-family: Roboto, Arial;
  font-weight: 500;
  line-height: 1.25;
  flex-direction: row;
  letter-spacing: 0.24px;
  justify-content: space-between;
}
.container-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  z-index: 1100;
  position: fixed;
}
.appBar-d1-0-2-4 {
  background-color: #ffffff;
  box-shadow: none;
  color: #363636;
}",
  <div>
    <header
      class=" container-0-2-1 container-d0-0-2-3 AppBarTop-gaya"
    >
      <div
        class="appBar-0-2-2 appBar-d1-0-2-4"
      >
        app bar
      </div>
    </header>
  </div>,
]
`;
